onnxscript/rewriter/ort_fusions/skip_normalization.py

```diff
@@ -2,115 +2,195 @@
-        skip_sum,
-        gamma,
-        beta,
-        axis=-1,
-        epsilon=epsilon,
-        stash_type=stash_type,
-    )
-    return normalized, skip_sum
-
-
-def _skip_layer_normalization(op, input, skip, gamma, beta, epsilon, stash_type):
-    if stash_type.value != 1:  # FLOAT type
-        return None
-    normalized, _mean, _inv_std_var, skip_sum = op.SkipLayerNormalization(
-        input,
-        skip,
-        gamma,
-        beta,
-        epsilon=epsilon,
-        _outputs=4,
-        _domain="com.microsoft",
-    )
-    return normalized, skip_sum
-
-
-# Fusion rule for Add + SkipLayerNormalization
-def _skip_layer_norm_add_bias_pattern(op, input, skip, gamma, beta, bias, epsilon, stash_type):
-    bias_sum = op.Add(input, bias)
-    normalized, _mean, _inv_std_var, skip_sum = op.SkipLayerNormalization(
-        bias_sum,
-        skip,
-        gamma,
-        beta,
-        epsilon=epsilon,
-        _outputs=4,
-        _domain="com.microsoft",
-    )
-    return normalized, skip_sum
-
-
-def _skip_layer_normalization_add_bias(
-    op, input, skip, gamma, beta, bias, epsilon, stash_type
-):
-    normalized, _mean, _inv_std_var, skip_sum = op.SkipLayerNormalization(
-        input,
-        skip,
-        gamma,
-        beta,
-        bias,
-        epsilon=epsilon,
-        _outputs=4,
-        _domain="com.microsoft",
-    )
-    return normalized, skip_sum
-
-
-_skip_layer_rule = pattern.RewriteRule(
-    _skip_layer_norm_pattern, _skip_layer_normalization, name="SkipLayerNorm"
+Dim = Union[int, ir.SymbolicDim]
+
+# Fusion rule for SkipRMSNormalization
+
+
+class SkipRmsNormFusion(pattern.RewriteRuleClassBase):
+    def __init__(self, name: str, has_bias: bool = False, bias_pre_add: bool = False):
+        """Fusion rule for SkipRMSNormalization."""
+        super().__init__(name=name)
+        self._has_bias = has_bias
+        self._bias_pre_add = bias_pre_add
+
+    def pattern(self, op, input, skip, gamma, bias, epsilon, stash_type):
+        if self._has_bias and self._bias_pre_add:
+            input = op.Add(input, bias)
+        skip_sum = op.Add(input, skip)
+        if self._has_bias and not self._bias_pre_add:
+            skip_sum = op.Add(skip_sum, bias)
+        normalized = op.SimplifiedLayerNormalization(
+            skip_sum,
+            gamma,
+            axis=-1,
+            epsilon=epsilon,
+            stash_type=stash_type,
+        )
+        return normalized, skip_sum
+
+    def check(self, op, input, skip, gamma, bias, epsilon, stash_type) -> pattern.MatchResult:  # type: ignore[name-defined]
+        """Check if the pattern matches conditions for use of SkipSimplifiedLayerNormalization op."""
+        check_result = pattern.MatchResult()
+        bindings: dict[str, Dim] = {}
+
+        def no_match(val: ir.Value, dims: Sequence[str]) -> bool:
+            return not _fusion_utils._check_shape(bindings, val, dims)
+
+        if no_match(input, ["B", "S", "D"]):
```

Comment:
For my understanding, how does no match work?

Reply:
What happens when the shapes do not take these names?

Reply:
The symbolic dim names like "B" are local to this check function ... it is used to ensure equivalence of different dims that are expected to be equivalent. "B" will be bound to the actual dimension, whether it is 1 or "BatchSize" or whatever.